File: tests/test_integration.py
"""Integration tests: Federated XGBoost with Shamir secret sharing."""

from typing import List

import numpy as np
import pandas as pd
import pytest

from privateboost import (
    Aggregator,
    BinConfiguration,
    Client,
    ShareHolder,
)

FEATURES = [
    "age",
    "sex",
    "cp",
    "trestbps",
    "chol",
    "fbs",
    "restecg",
    "thalach",
    "exang",
    "oldpeak",
    "slope",
    "ca",
    "thal",
]

COLUMNS = [*FEATURES, "target"]


@pytest.fixture(scope="session")
def heart_disease_df(tmp_path_factory: pytest.TempPathFactory) -> pd.DataFrame:
    """Load UCI Heart Disease dataset with caching."""
    cache_dir = tmp_path_factory.mktemp("data")
    cache_file = cache_dir / "heart_disease.csv"

    if cache_file.exists():
        df = pd.read_csv(cache_file)
    else:
        url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"
        df = pd.read_csv(url, names=COLUMNS, na_values="?")
        df = df.dropna()
        df["target"] = (df["target"] > 0).astype(int)
        df.to_csv(cache_file, index=False)

    return df


def create_clients(
    df: pd.DataFrame,
    features: List[str],
    shareholders: List[ShareHolder],
    threshold: int = 2,
) -> List[Client]:
    """Create Client instances from DataFrame rows."""
    return [
        Client(
            client_id=f"client_{idx}",
            features=row[features].values,
            target=float(row["target"]),
            shareholders=shareholders,
            threshold=threshold,
        )
        for idx, row in df.iterrows()
    ]


def compute_initial_statistics(
    clients: List[Client],
    aggregator: Aggregator,
) -> tuple[List[BinConfiguration], float]:
    """Compute bins and initial prediction from client statistics."""
    for client in clients:
        client.submit_stats()

    bins = aggregator.define_bins()
    initial_pred = float(aggregator.means[-1])

    return bins, initial_pred


def train_tree(
    clients: List[Client],
    aggregator: Aggregator,
    bins: List[BinConfiguration],
    max_depth: int,
    round_id: int,
) -> None:
    """Train a single tree."""
    for depth in range(max_depth):
        for client in clients:
            client.submit_gradients(
                bins,
                aggregator.model,
                aggregator.splits,
                round_id=round_id,
                depth=depth,
                loss="logistic",
            )

        if not aggregator.compute_splits(depth=depth, min_child_weight=1.0):
            break

    aggregator.finish_round()


def test_xgboost_heart_disease_shamir(heart_disease_df: pd.DataFrame):
    """End-to-end test: federated XGBoost with Shamir 2-of-3 achieves >75% accuracy."""
    np.random.seed(42)

    df = heart_disease_df.sample(frac=1, random_state=42).reset_index(drop=True)
    split_idx = int(len(df) * 0.8)
    df_train = df.iloc[:split_idx]
    df_test = df.iloc[split_idx:]

    threshold = 2
    min_clients = 10
    learning_rate = 0.15
    lambda_reg = 2.0

    shareholders = [
        ShareHolder(party_id=i, x_coord=i + 1, min_clients=min_clients) for i in range(3)
    ]
    aggregator = Aggregator(
        shareholders,
        n_bins=10,
        threshold=threshold,
        min_clients=min_clients,
        learning_rate=learning_rate,
        lambda_reg=lambda_reg,
    )

    train_clients = create_clients(df_train, FEATURES, shareholders, threshold=threshold)

    bins, initial_pred = compute_initial_statistics(train_clients, aggregator)

    print(f"Number of training clients: {aggregator.n_clients}")
    print(f"Initial prediction (target mean): {initial_pred:.4f}")

    n_trees = 15
    max_depth = 3

    for round_id in range(n_trees):
        train_tree(
            train_clients,
            aggregator,
            bins,
            max_depth,
            round_id,
        )

    test_features_arr = df_test[FEATURES].values.astype(float)
    test_targets = df_test["target"].values

    test_preds = aggregator.model.predict(test_features_arr)
    test_classes = (test_preds >= 0.5).astype(int)
    test_accuracy = np.mean(test_classes == test_targets)

    print(f"Test accuracy: {test_accuracy:.2%}")
    assert test_accuracy > 0.75, f"Expected >75% accuracy, got {test_accuracy:.2%}"


def test_xgboost_heart_disease_path_hiding(heart_disease_df: pd.DataFrame):
    """End-to-end test: path hiding produces same accuracy as without."""
    np.random.seed(42)

    df = heart_disease_df.sample(frac=1, random_state=42).reset_index(drop=True)
    split_idx = int(len(df) * 0.8)
    df_train = df.iloc[:split_idx]
    df_test = df.iloc[split_idx:]

    threshold = 2
    min_clients = 10
    learning_rate = 0.15
    lambda_reg = 2.0

    shareholders = [
        ShareHolder(party_id=i, x_coord=i + 1, min_clients=min_clients) for i in range(3)
    ]
    aggregator = Aggregator(
        shareholders,
        n_bins=10,
        threshold=threshold,
        min_clients=min_clients,
        learning_rate=learning_rate,
        lambda_reg=lambda_reg,
    )

    train_clients = create_clients(df_train, FEATURES, shareholders, threshold=threshold)

    bins, _ = compute_initial_statistics(train_clients, aggregator)

    n_trees = 15
    max_depth = 3

    for round_id in range(n_trees):
        for depth in range(max_depth):
            for client in train_clients:
                client.submit_gradients(
                    bins,
                    aggregator.model,
                    aggregator.splits,
                    round_id=round_id,
                    depth=depth,
                    loss="logistic",
                    hide_path=True,
                )

            if not aggregator.compute_splits(depth=depth, min_child_weight=1.0):
                break

        aggregator.finish_round()

    test_features_arr = df_test[FEATURES].values.astype(float)
    test_targets = df_test["target"].values

    test_preds = aggregator.model.predict(test_features_arr)
    test_classes = (test_preds >= 0.5).astype(int)
    test_accuracy = np.mean(test_classes == test_targets)

    print(f"Test accuracy (path hiding): {test_accuracy:.2%}")
    assert test_accuracy > 0.75, f"Expected >75% accuracy with path hiding, got {test_accuracy:.2%}"


def test_min_clients_enforcement():
    """Test that shareholders enforce minimum client threshold."""
    min_clients = 10
    shareholders = [
        ShareHolder(party_id=i, x_coord=i + 1, min_clients=min_clients) for i in range(3)
    ]

    clients = [Client(f"c{i}", [1.0, 2.0], 0.0, shareholders, threshold=2) for i in range(5)]

    for client in clients:
        client.submit_stats()

    commitments = list(shareholders[0].get_stats_commitments())

    with pytest.raises(ValueError, match="minimum"):
        shareholders[0].get_stats_sum(commitments=commitments)

File: rust/privateboost/tests/test_integration.rs
use privateboost::*;
use std::path::Path;

#[test]
fn test_xgboost_heart_disease_shamir() {
    // CSV is pre-shuffled with numpy random_state=42 to match Python test split
    let dataset = read_csv(Path::new("tests/data/heart_disease.csv"), "target").unwrap();

    let features = dataset.features;
    let targets = dataset.targets;

    let split_idx = (features.len() as f64 * 0.8) as usize;
    let train_features = &features[..split_idx];
    let train_targets = &targets[..split_idx];
    let test_features = &features[split_idx..];
    let test_targets = &targets[split_idx..];

    let threshold = 2;
    let min_clients = 10;
    let learning_rate = 0.15;
    let lambda_reg = 2.0;
    let n_bins = 10;
    let n_shareholders = 3;
    let n_trees = 15;
    let max_depth = 3;

    let shareholders: Vec<ShareHolder> = (0..n_shareholders)
        .map(|i| ShareHolder::new(i, (i + 1) as u64, min_clients))
        .collect();

    let mut aggregator = Aggregator::new(
        shareholders,
        n_bins,
        threshold,
        min_clients,
        learning_rate,
        lambda_reg,
        BinMethod::Gaussian,
    )
    .unwrap();

    let mut clients: Vec<Client> = train_features
        .iter()
        .zip(train_targets.iter())
        .enumerate()
        .map(|(idx, (feats, &target))| {
            Client::new(
                format!("client_{}", idx),
                feats.clone(),
                target,
                n_shareholders,
                threshold,
                None,
            )
        })
        .collect();

    // Phase 1: Statistics
    for client in &mut clients {
        let shares = client.compute_stat_shares().unwrap();
        for (i, share) in shares.into_iter().enumerate() {
            aggregator.shareholders_mut()[i].receive_stats(share);
        }
    }
    aggregator.define_bins().unwrap();

    println!("Training clients: {}", aggregator.n_clients());
    println!(
        "Initial prediction (target mean): {:.4}",
        aggregator.feature_stats().unwrap().last().unwrap().mean
    );

    // Phase 2: Tree training
    for _ in 0..n_trees {
        for _ in 0..max_depth {
            let ctx = aggregator.round_context();
            for client in &mut clients {
                let shares = client
                    .compute_gradient_shares(&ctx, &Loss::Logistic, false)
                    .unwrap();
                for (i, share) in shares.into_iter().enumerate() {
                    aggregator.shareholders_mut()[i].receive_gradients(share);
                }
            }
            if !aggregator.compute_splits(0.0, 1.0).unwrap() {
                break;
            }
        }
        aggregator.finish_round();
    }

    // Phase 3: Prediction
    let test_features_vec: Vec<Vec<f64>> = test_features.to_vec();
    let predictions = aggregator.model().predict(&test_features_vec);

    // Threshold at 0.5 on raw log-odds to match Python reference behavior
    let correct: usize = predictions
        .iter()
        .zip(test_targets.iter())
        .filter(|&(&pred, &target)| {
            let class = if pred >= 0.5 { 1.0 } else { 0.0 };
            (class - target).abs() < 1e-10
        })
        .count();

    let accuracy = correct as f64 / test_targets.len() as f64;
    println!("Test accuracy: {:.2}%", accuracy * 100.0);
    assert!(
        accuracy >= 0.75,
        "Expected >=75% accuracy, got {:.2}%",
        accuracy * 100.0
    );
}

#[test]
fn test_xgboost_heart_disease_path_hiding() {
    let dataset = read_csv(Path::new("tests/data/heart_disease.csv"), "target").unwrap();
    let features = dataset.features;
    let targets = dataset.targets;

    let split_idx = (features.len() as f64 * 0.8) as usize;
    let train_features = &features[..split_idx];
    let train_targets = &targets[..split_idx];
    let test_features = &features[split_idx..];
    let test_targets = &targets[split_idx..];

    let threshold = 2;
    let min_clients = 10;
    let n_shareholders = 3;
    let n_trees = 15;
    let max_depth = 3;

    let shareholders: Vec<ShareHolder> = (0..n_shareholders)
        .map(|i| ShareHolder::new(i, (i + 1) as u64, min_clients))
        .collect();

    let mut aggregator = Aggregator::new(shareholders, 10, threshold, min_clients, 0.15, 2.0, BinMethod::Gaussian).unwrap();

    let mut clients: Vec<Client> = train_features
        .iter()
        .zip(train_targets.iter())
        .enumerate()
        .map(|(idx, (feats, &target))| {
            Client::new(format!("client_{}", idx), feats.clone(), target, n_shareholders, threshold, None)
        })
        .collect();

    for client in &mut clients {
        let shares = client.compute_stat_shares().unwrap();
        for (i, share) in shares.into_iter().enumerate() {
            aggregator.shareholders_mut()[i].receive_stats(share);
        }
    }
    aggregator.define_bins().unwrap();

    for _ in 0..n_trees {
        for _ in 0..max_depth {
            let ctx = aggregator.round_context();
            for client in &mut clients {
                let shares = client
                    .compute_gradient_shares(&ctx, &Loss::Logistic, true)
                    .unwrap();
                for share in shares {
                    let sh_idx = share.share.x as usize - 1;
                    aggregator.shareholders_mut()[sh_idx].receive_gradients(share);
                }
            }
            if !aggregator.compute_splits(0.0, 1.0).unwrap() {
                break;
            }
        }
        aggregator.finish_round();
    }

    let predictions = aggregator.model().predict(&test_features.to_vec());
    let correct: usize = predictions
        .iter()
        .zip(test_targets.iter())
        .filter(|&(&pred, &target)| {
            let class = if pred >= 0.5 { 1.0 } else { 0.0 };
            (class - target).abs() < 1e-10
        })
        .count();
    let accuracy = correct as f64 / test_targets.len() as f64;
    println!("Path hiding accuracy: {:.2}%", accuracy * 100.0);
    assert!(accuracy >= 0.75, "Expected >=75% with path hiding, got {:.2}%", accuracy * 100.0);
}

#[test]
fn test_min_clients_enforcement() {
    let min_clients = 10;
    let n_shareholders = 3;
    let threshold = 2;

    let shareholders: Vec<ShareHolder> = (0..n_shareholders)
        .map(|i| ShareHolder::new(i, (i + 1) as u64, min_clients))
        .collect();

    let mut aggregator =
        Aggregator::new(shareholders, 10, threshold, min_clients, 0.1, 1.0, BinMethod::Gaussian).unwrap();

    let mut clients: Vec<Client> = (0..5)
        .map(|i| {
            Client::new(
                format!("c{}", i),
                vec![1.0, 2.0],
                0.0,
                n_shareholders,
                threshold,
                None,
            )
        })
        .collect();

    for client in &mut clients {
        let shares = client.compute_stat_shares().unwrap();
        for (i, share) in shares.into_iter().enumerate() {
            aggregator.shareholders_mut()[i].receive_stats(share);
        }
    }

    let result = aggregator.define_bins();
    assert!(result.is_err());
}

#[test]
fn test_gaussian_binning_edges_denser_at_center() {
    let n = 200;
    // Linearly spaced data from -3 to 3 (uniform, but mean≈0, std≈1.73)
    let features: Vec<Vec<f64>> = (0..n)
        .map(|i| vec![-3.0 + 6.0 * (i as f64) / (n as f64 - 1.0)])
        .collect();
    let targets: Vec<f64> = (0..n).map(|i| if i % 2 == 0 { 1.0 } else { 0.0 }).collect();

    let n_shareholders = 3;
    let threshold = 2;
    let min_clients = 5;
    let n_bins = 10;

    let shareholders: Vec<ShareHolder> = (0..n_shareholders)
        .map(|i| ShareHolder::new(i, (i + 1) as u64, min_clients))
        .collect();

    let mut aggregator = Aggregator::new(
        shareholders, n_bins, threshold, min_clients, 0.1, 1.0, BinMethod::Gaussian,
    ).unwrap();

    let mut clients: Vec<Client> = features.iter().zip(targets.iter()).enumerate()
        .map(|(i, (f, &t))| Client::new(format!("c{i}"), f.clone(), t, n_shareholders, threshold, None))
        .collect();

    for client in &mut clients {
        let shares = client.compute_stat_shares().unwrap();
        for (i, share) in shares.into_iter().enumerate() {
            aggregator.shareholders_mut()[i].receive_stats(share);
        }
    }

    let configs = aggregator.define_bins().unwrap();
    let config = &configs[0];

    // Gaussian quantile edges should be more concentrated near the center
    let inner = &config.edges[1..config.edges.len() - 1];
    let mid = inner.len() / 2;
    let center_gap = (inner[mid] - inner[mid - 1]).abs();
    let outer_gap = (inner[1] - inner[0]).abs();
    assert!(
        center_gap < outer_gap,
        "Gaussian edges should be denser near center: center_gap={center_gap}, outer_gap={outer_gap}"
    );
}
